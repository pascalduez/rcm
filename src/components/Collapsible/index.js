// @flow

export { default as Collapsible } from './Collapsible';
export type { Props as CollapsibleProps } from './Collapsible';
