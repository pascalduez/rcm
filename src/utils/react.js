// @flow

import type { Element } from 'react';

export const getElementDisplayName = (element: Element<*>): string =>
  element.type.displayName ||
  // Function name.
  element.type.name ||
  // Function without a name.
  (typeof element.type === 'function' ? 'No Display Name' : element.type);
