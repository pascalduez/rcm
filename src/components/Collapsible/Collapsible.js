// @flow

import * as React from 'react';
import { canUseDOM } from '../../utils';

type State = {
  collapsed: boolean,
};

export type Props = {
  collapsed?: boolean,
  render: (
    getActionProps: (*) => {},
    getDrawerProps: (*) => {},
    isCollapsed: boolean
  ) => React.Element<*>,
};

let idcounter = 0;

class Collapsible extends React.Component<Props, State> {
  id: string;

  static defaultProps = {
    collapsed: false,
  };

  state = {
    collapsed: this.props.collapsed || false,
  };

  /**
   * Unique Id generation.
   * All SVGs on the page should have a unique ID.
   * Used for `title`, `desc` a11y.
   * Saddly not work with SSR.
   * See https://git.io/vdoUM
   */
  componentWillMount() {
    this.id = `collapsible-${(idcounter += 1)}`;
  }

  /**
   * Allow to update the `collapsed` state via props.
   */
  componentWillReceiveProps(nextProps: Props) {
    if (this.props.collapsed !== nextProps.collapsed) {
      this.setState({
        collapsed: nextProps.collapsed,
      });
    }
  }

  /**
   * Handle click events on the Action.
   */
  handleClick = (evt: SyntheticEvent<HTMLButtonElement>): void => {
    evt.preventDefault();

    this.setState(prevState => ({
      collapsed: !prevState.collapsed,
    }));
  };

  // $FlowFixMe Destructuring default params + Flow annotation breaks react-docgen.
  getActionProps = ({ style, onClick, ...rest } = {}) => {
    const handleClick = (evt: SyntheticEvent<HTMLButtonElement>) => {
      this.handleClick(evt);
      if (onClick) onClick(evt);
    };

    const localStyle = canUseDOM() ? {} : { display: 'none' };

    return {
      type: 'button',
      'aria-expanded': !this.state.collapsed,
      'aria-controls': this.id,
      children: this.state.collapsed ? 'Expand' : 'Collapse',
      style: { ...style, ...localStyle },
      onClick: handleClick,
      ...rest,
    };
  };

  // $FlowFixMe Destructuring default params + Flow annotation breaks react-docgen.
  getDrawerProps = ({ style, ...rest } = {}) => ({
    'aria-hidden': this.state.collapsed,
    id: this.id,
    style: {
      display: this.state.collapsed ? 'none' : 'block',
      ...style,
    },
    ...rest,
  });

  render() {
    return this.props.render(
      this.getActionProps,
      this.getDrawerProps,
      this.state.collapsed
    );
  }
}

export default Collapsible;
