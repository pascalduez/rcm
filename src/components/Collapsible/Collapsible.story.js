// @flow

import React from 'react';
import { storiesOf } from '@storybook/react';

import Collapsible from './Collapsible';

storiesOf('Collapsible', module).add('Demo', () => ());
