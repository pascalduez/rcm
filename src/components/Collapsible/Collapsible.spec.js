// @flow

import * as React from 'react';
import { render, mount } from 'enzyme';

import Collapsible from './Collapsible';

const TestComponent = (
  <Collapsible
    render={(getActionProps, getDrawerProps) => (
      <div>
        <button {...getActionProps()}>Test</button>
        <div {...getDrawerProps()}>Test text...</div>
      </div>
    )}
  />
);

describe('<Collapsible />', () => {
  it('should render the passed elements', () => {
    const wrapper = render(TestComponent);

    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle the Drawer style', () => {
    const wrapper = mount(TestComponent);

    expect(wrapper.find('div').at(1)).toHaveStyle('display', 'block');

    wrapper.find('button').simulate('click');

    expect(wrapper.find('div').at(1)).toHaveStyle('display', 'none');
  });

  describe('the `collapsed` property', () => {
    it('should trigger a state update', () => {
      const wrapper = mount(TestComponent);

      expect(wrapper).toHaveState('collapsed', false);

      wrapper.setProps({ collapsed: true });

      expect(wrapper).toHaveState('collapsed', true);
    });
  });

  describe('the `getActionProps` helper', () => {
    it('should properly merge properties', () => {
      const spy = jest.fn();
      const wrapper = mount(
        <Collapsible
          render={getActionProps => (
            <button
              {...getActionProps({
                onClick: spy,
              })}
            />
          )}
        />
      );

      wrapper.find('button').simulate('click');

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('the `getDrawerProps` helper', () => {
    it('should properly merge properties', () => {
      const wrapper = mount(
        <Collapsible
          render={(getActionProps, getDrawerProps) => (
            <div
              {...getDrawerProps({
                style: { minHeight: '200px' },
              })}
            >
              Test text...
            </div>
          )}
        />
      );

      expect(wrapper.find('div')).toHaveStyle('display', 'block');
      expect(wrapper.find('div')).toHaveStyle('minHeight', '200px');

      wrapper.setProps({ collapsed: true });

      expect(wrapper.find('div')).toHaveStyle('display', 'none');
      expect(wrapper.find('div')).toHaveStyle('minHeight', '200px');
    });
  });

  describe('the `isCollapsed` helper', () => {
    it('should properly reflect current state', () => {
      const spy = jest.fn(() => <div>Test text...</div>);
      const wrapper = mount(<Collapsible render={spy} />);

      expect(spy).toHaveBeenCalledWith(
        expect.anything(),
        expect.anything(),
        false
      );

      wrapper.setProps({ collapsed: true });

      expect(spy).toHaveBeenCalledWith(
        expect.anything(),
        expect.anything(),
        true
      );
    });
  });
});
