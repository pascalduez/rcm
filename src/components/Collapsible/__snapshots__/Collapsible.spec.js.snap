// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapsible /> should render the passed elements 1`] = `
<div>
  <button
    aria-controls="collapsible-1"
    aria-expanded="true"
    type="button"
  >
    Test
  </button>
  <div
    aria-hidden="false"
    id="collapsible-1"
    style="display:block"
  >
    Test text...
  </div>
</div>
`;
